### ISSUE TITLE
Excessive queries on Oracle can cause connection pool to get exhausted and never released

### ISSUE THREAD

**Issue by flamber** [metabase/metabase#15817]
**Describe the bug**
Under specific condition, it's possible to "overload" the connection pool to Oracle, which then leaves the connections on Oracle in `INACTIVE` status, but then Metabase cannot reuse those connections and will stop being able to query the database.
~Maybe caused by Metabase using it's own connection pool c3p0, but there's also a connection pool in Oracle driver too - UCP.~

**Possible temporary remedy/workaround**
Have a cronjob running every minute with the following query on the database to kill old, inactive connections.
It can be tweaked to be only connections from Metabase, see the columns available in `select * from v$session`
```
BEGIN
  FOR sess IN (select sid,serial# from v$session where status='INACTIVE')
  LOOP
    EXECUTE IMMEDIATE 'alter system kill session ''' || sess.sid  || ',' || sess.serial# || ''' immediate';
  END LOOP;
END;
```

**To Reproduce**
1. Setup Oracle database
```
docker run --name oracle11g -d -p 1521:1521 -p 8080:8080 -e ORACLE_ALLOW_REMOTE=true wnameless/oracle-xe-11g-r2
```
2. Connect to Oracle in Metabase - host: `localhost` port: `1521` sid: `xe` username: `system` password: `oracle`
3. Create 20 unique long-running (60 seconds) questions - it's very important that each is an unique query, since it will otherwise add the queries to "queued" instead of "Queries in flight", which is what is need to reproduce the problem.
```
Q1: select 01, dbms_pipe.receive_message(('a'),60), coalesce(null, null [[,{{filter}}]]) as filter from dual
Q2: select 02, dbms_pipe.receive_message(('a'),60), coalesce(null, null [[,{{filter}}]]) as filter from dual
Q3: ...
```
4. Add all questions to a dashboard (connect a "Other categories" filter for more real-world scenario, since changing the filter while queries are in-flight is more likely than leaving the dashboard).
5. In Firefox change `network.http.max-persistent-connections-per-server` in about:config to `25` to simulate http/2
6. Now go and visit the dashboard - when everything is still loading (queries in flight), leave the dashboard, which should then result in a lot of lines like this, where it says "Queries in flight: 0" at the end.
```
POST /api/card/512/query 202 [ASYNC: canceled] 7.1 s (5 DB calls) App DB connections: 0/10 Jetty threads: 3/50 (5 idle, 0 queued) (82 total active threads) Queries in flight: 0 (0 queued)
```
7. Connect to Oracle server via non-Metabase with user: `sys` password: `oracle` role: SYSDBA and run this query to see connections, which will show 15 `ACTIVE` connections:
`SELECT sid, status, sql_exec_start, prev_exec_start, logon_time, program FROM v$session WHERE program LIKE 'MB v%';`
8. Even after the query running period (60 seconds), the only change will be that the connections on Oracle will change to `INACTIVE`, but they will not be released after 5/15 minutes (c3p0 timeout), and it's now not possible to make any queries in Metabase against the Oracle database - even a simple `select 1 from dual` will not run until Metabase has been restarted or the connections on Oracle has been forced-closed.
9. Going to the dashboard and changing filter while queries are running, will just cause the "Queries in flight" counter to go up.

**Expected behavior**
Same behavior as other drivers.

**Information about your Metabase Installation:**
Tested 0.38.4 and 0.39.0, but it's likely been a problem since at least 0.35.0
Related #15774

**Comment by jeff303** [comment]
I am unable to reproduce this on 40.  I followed the exact steps outlined (taking a few shortcuts to insert questions and associate to the dashboard via SQL INSERT statements, to save some time).  Through Oracle, I can see all connections go from `ACTIVE` to `INACTIVE` within a couple minutes, as expected.  If I then immediately run a native query like `SELECT 42 FROM dual`, then it succeeds with no issue.

Also, UCP should not be involved at all unless people are bundling the `ucp.jar` file along with `ojdbc8.jar` (and we [don't seem to be telling them to](https://www.metabase.com/docs/latest/administration-guide/databases/oracle.html)), and the code calls the `PoolDataSourceFactory` code as outlined [here](https://docs.oracle.com/cd/E11882_01/java.112/e12265/connect.htm#CHDGEJJF).

@flamber , we may need to set up some time to screen share on this one.

**Comment by flamber** [comment]
@jeff303 I can still reproduce on `c040f15` (2021-06-25) with 20 unique "sleep 60" cards on a dashboard, which I then load, then close the browser tab (do that a couple of times, just to stress the system). The Oracle just have 15 connections in `waiting idle` state, but no new queries can be made - no matter if I wait 5 minutes (was previously 15 minutes).

The connection pool info just shows this after 20 minutes, when I have no more tabs open:
```
{"connection-pools":{"metabase-h2-app-db":{"numConnections":15,"numIdleConnections":15,"numBusyConnections":0,"minPoolSize":1,"maxPoolSize":15},"db-29-oracle-null":{"numConnections":15,"numIdleConnections":0,"numBusyConnections":15,"minPoolSize":1,"maxPoolSize":15}}}
```
And the last log line shows only 2 queries in flight, which I don't know which those could be:
```
2021-06-29 17:13:48,348 DEBUG middleware.log :: GET /api/util/bug_report_details 200 29.7 ms (1 DB calls) App DB connections: 0/15 Jetty threads: 4/50 (1 idle, 0 queued) (160 total active threads) Queries in flight: 2 (0 queued)
```
I can make queries to any other database besides the Oracle ID 29.

Unless you're testing on HTTP/2, then you need to simulate that in the browser (Firefox `network.http.max-persistent-connections-per-server=100` and `network.http.max-urgent-start-excessive-connections-per-host=100`)

I'm testing with ojdbc 19.3 (because that is what we recommend and distribute in EE) and Oracle 11g (because that was the only thing available until very recently).

**Comment by jeff303** [comment]
@flamber , what is the exact jar name you're using for the Oracle JDBC driver?  I tried with `ojdbc8.jar` (downloaded from that site you linked).

**Comment by flamber** [comment]
@jeff303 If you're using EE, then it's included already, but for OSS, then I just add ojdbc8.jar to plugins.
This issue was originally observed on Cloud, so that was running EE with the included 19.3, but I don't know which version of Oracle database it was.

**Comment by jeff303** [comment]
Tried again today to reproduce, and this time I couldn't.  😢 

At least I wrote a Postgres snippet to populate all the questions, since that setup is always kind of time consuming whenever I try to tackle this issue once more.  It will insert the 20 cards and add them to a dashboard.


```

DO $$
DECLARE 
  -- this assumes there is only one Oracle DB created in this MB instance
  db_id INTEGER := (SELECT id FROM METABASE_DATABASE WHERE engine = 'oracle');
  -- find some active user to own the cards
  creator_id INTEGER := (SELECT id FROM core_user WHERE is_active AND last_login IS NOT NULL);
  -- set the collection_id that the cards will be added to (create manually)
  collection_id INTEGER := 12;
  -- set the dashboard_id that the cards will be added to (create manually)
  dashboard_id INTEGER := 2;
BEGIN
FOR i IN 1..20 LOOP

  WITH inserted_card AS (
   INSERT INTO report_card
    (
      created_at,
      updated_at,
      name,
      description,
      display,
      dataset_query,
      visualization_settings,
      creator_id,
      database_id,
      table_id,
      query_type,
      archived,
      collection_id,
      public_uuid,
      made_public_by_id,
      enable_embedding,
      embedding_params,
      cache_ttl,
      result_metadata,
      collection_position
    )
    VALUES
    (
      CURRENT_TIMESTAMP,
      CURRENT_TIMESTAMP,
      CONCAT('Oracle Q',i),
      NULL,
      'table',
      CONCAT('{"type":"native","native":{"query":"select 0',i,', dbms_pipe.receive_message((''a''),60), coalesce(null, null [[,{{filter}}]]) as filter from dual\n","template-tags":{"filter":{"id":"38d119cf-287c-3e3e-c8da-f4cd39ffaf98","name":"filter","display-name":"Filter","type":"text"}}},"database":',db_id,'}'),
      '{}',
      creator_id,
      db_id,
      NULL,
      'native',
      FALSE,
      collection_id,
      NULL,
      NULL,
      FALSE,
      NULL,
      NULL,
      '[{"name":"01","display_name":"01","base_type":"type/Decimal","field_ref":["field","01",{"base-type":"type/Decimal"}],"semantic_type":null,"fingerprint":{"global":{"distinct-count":1,"nil%":0.0},"type":{"type/Number":{"min":1.0,"q1":1.0,"q3":1.0,"max":1.0,"sd":null,"avg":1.0}}}},{"name":"DBMS_PIPE.RECEIVE_MESSAGE((''A''),60)","display_name":"DBMS_PIPE.RECEIVE_MESSAGE((''A''),60)","base_type":"type/Decimal","field_ref":["field","DBMS_PIPE.RECEIVE_MESSAGE((''A''),60)",{"base-type":"type/Decimal"}],"semantic_type":null,"fingerprint":{"global":{"distinct-count":1,"nil%":0.0},"type":{"type/Number":{"min":1.0,"q1":1.0,"q3":1.0,"max":1.0,"sd":null,"avg":1.0}}}},{"name":"FILTER","display_name":"FILTER","base_type":"type/Text","field_ref":["field","FILTER",{"base-type":"type/Text"}],"semantic_type":null,"fingerprint":{"global":{"distinct-count":1,"nil%":1.0},"type":{"type/Text":{"percent-json":0.0,"percent-url":0.0,"percent-email":0.0,"percent-state":0.0,"average-length":0.0}}}}]',
      NULL
    ) RETURNING id
  )
  
  INSERT INTO report_dashboardcard (created_at,updated_at,"sizeX","sizeY",row,col,card_id,dashboard_id,parameter_mappings,visualization_settings) 
  VALUES (CURRENT_TIMESTAMP,CURRENT_TIMESTAMP,4,4,((i-1)/4),4*((i-1)%4),(select id from inserted_card),dashboard_id,'[]','{}');
  
END LOOP;
END;
$$;
```
